go/pkg/tool/tool_test.go

```diff
@@ -276,3 +280,50 @@ func TestTool_UploadBlob(t *testing.T) {
 		t.Fatalf("Expected 1 write for blob '%v', got %v", dg.String(), cas.BlobWrites(dg))
 	}
 }
+
+func TestTool_UploadTree(t *testing.T) {
+	tempDir := path.Join(t.TempDir(), "TestTool_UploadTree")
+	os.MkdirAll(tempDir, os.ModePerm)
+
+	e, cleanup := fakes.NewTestEnv(t)
+	defer cleanup()
+	cas := e.Server.CAS
+
+	files := []string{
+		"blob",
+		"foo/blob",
+		"foo/bar/blob",
+		"bar/blob",
+	}
+
+	for _, file := range files {
+		path := path.Join(tempDir, file)
+		os.MkdirAll(filepath.Dir(path), os.ModePerm)
+		if err := ioutil.WriteFile(path, []byte(file), 0777); err != nil {
+			t.Fatalf("Could not create temp blob '%v': %v", file, err)
+		}
+	}
+
+	tmpFile := path.Join(tempDir, "blob")
+	if err := ioutil.WriteFile(tmpFile, []byte("Hello, World!"), 0777); err != nil {
+		t.Fatalf("Could not create temp blob: %v", err)
+	}
+
+	toolClient := &Client{GrpcClient: e.Client.GrpcClient}
+	if err := toolClient.UploadTree(context.Background(), tempDir); err != nil {
+		t.Fatalf("UploadTree('%v') failed: %v", tmpFile, err)
+	}
+
+	// First request should upload the blob.
+	if cas.WriteReqs() != 4 {
+		t.Fatalf("Expected 4 writes, got %v", cas.WriteReqs())
+	}
+
+	// Retries should check whether the blob already exists and skip uploading if it does.
+	if err := toolClient.UploadTree(context.Background(), tempDir); err != nil {
+		t.Fatalf("UploadTree('%v') failed: %v", tmpFile, err)
+	}
+	if cas.WriteReqs() != 4 {
```

Comment:
If we are skipping the blob upload during second call to the same `tempDir` path, then shouldn't we have 0 write requests?

Reply:
The stats aren't reset, so they continue to be 4 from the initial upload. If the second call uploads anything we'd get > 4 here.